### ISSUE TITLE
 I can not use the geometry tool with the shortcut pan (Space) at the same time

### ISSUE THREAD

**Issue by tghart** [opentoonz/opentoonz#1235]
Hello, i have problem in opentoonz.

I user Pentab Wacom. When i drawing with geometry tool i can't use shorcut pan (space) at the same time. but when i use mouse and use middle button at the same time it's working.

Example Video : 
https://drive.google.com/open?id=0ByaxsBi8eeBDRFhZX2ZzR3hoWmc

Please Fix It
Thanks You

**Comment by turtletooth** [comment]
Right now the spacebar is the shortcut for the hand tool. Middle click does pan movement also. The only option I can think of is to hard code the spacebar to also do what the middle click does. Right now- the hand tool is a full separate tool and not just a temporary navigation button.

Hard coding space bar to do the same as the hand tool would work but the downside is that you couldn't change the shortcut.

**Comment by turtletooth** [comment]
from #1236 

If this is more of a feature request than a bug, please let me know and I'll repost it on Google Groups.

While you're editing a stroke with the Control Point Editor Tool, for example, you may want to rotate, zoom or pan the view, and when you do that the selection is cleared.

![opentoonz - navigation deselect bug](https://cloud.githubusercontent.com/assets/28221053/26519987/b52f76b0-42b9-11e7-9ed9-828c0202c83c.gif)

Expected behaviour:
Navigating, especially in this "override" mode (when the previous tool is restored afterwards) should preserve your selection so a user doesn't have to reselect the stroke each time they navigate.

**Comment by beeheemooth** [comment]
> The only option I can think of is to hard code the spacebar to also do what the middle click does. 

> 
I support your decision! The spacebar, in my opinion, should only temporarily change pan movement, I wrote about this earlier. #1100

**Comment by turtletooth** [comment]
It probably wouldn't be hard to change in the code, I just need to wait and here from other users what they think is best.

**Comment by RNavega** [comment]
The geometry tool case seems to be "modal", the line and handles are only drawn while the tool is active. 
When you (temporarily) switch to navigate, your unfinished drawing gets canceled.
https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/tnztools/geometrictool.cpp#L531

With the control point editor it's similar: the control points are only drawn while the tool is active, and the active stroke is forgotten when you get back to this tool after navigating.
1) You press the keys to invoke the temporary tool. The previous tool is stored:
https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/toonz/sceneviewerevents.cpp#L635-L654

2) The tool is switched. When it's deactivated, the control point editor tool doesn't draw the points and handles (but the stroke is still remembered)
https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/tnztools/toolhandle.cpp#L43
https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/tnztools/controlpointeditortool.cpp#L854

3) When you release the navigation keys and the control point editor tool is restored, its onActivated function forgets what was the last stroke:
https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/tnztools/controlpointeditortool.cpp#L848

So some solution would probably involve calling the ::draw of control point editor, and to not use ::setStroke( null, -1 ) when it's activated from a tool restore.

**Comment by RodneyBaker** [comment]
No updates on this issue since May 2017.
Recommend reviewing the issue.

**Comment by manongjohn** [comment]
In relation to #1100 and this one, I believe I've discovered why holding a tool shortcut for temporary access to that tool doesn't work in most cases.

Holding a letter-only shortcut causes that held key to be repeated.  Think about writing a document and holding the letter 'A'.  It will repeat  writing 'A' until you let go.

Because of this automatic repeat, it is quickly creating alternating ShortcutOverride and KeyRelease events, essentially the same effect as hitting the key repeatedly.  This prevents OT from determining if the key is actually being held for a long enough period to enable that temporary access feature.

However, if you do a ctrl/space+letter shortcut, it does work.  This is because the ctrl/space does not cause the key to be automatically repeated and thus a KeyRelease event is not actually generated until you stop.

A user workaround in the meantime would be to change the Hand Tool shortcut to include a ctrl/space. Since ctrl-space=rotation and shift-space=zoom, you'd need to change the Hand Tool to ctrl/space+some other letter.

Still trying to figure out how to fix this (Other developer suggestions welcome of course).

**Comment by beeheemooth** [comment]
There is one more thing that concerns this problem.

The order of keystrokes to switch to navigation mode. For example, to rotate the canvas, you must press the "Shift" and "Space" buttons in the following order: "Shift + Space" and not "Space + Shift", otherwise the tool will not turn on. In other programs, such as ClipStudio, the order of keystrokes doesn't matter. The user doesn't think about it - he simply presses two buttons at the same time.

**Comment by RodneyBaker** [comment]
@manongjohn 
It seems to me we need a pre-check that first establishes where we are in OpenToonz in order to assess what the user is desiring for the text-type keypress.  There are two types of check I can think of 1) Explicitly check to see if we are already in a 'mode' that using standard typing such as would be used for the Type Tool, NoteLevels or Notes 2) (The reverse) Check to make we are not in those text modes.

**Comment by Orphanlast** [comment]
Panning by middle mouse clicks works

**Comment by shun-iwasawa** [comment]
@manongjohn Good catch. May using `bool QKeyEvent::isAutoRepeat () const` and ignoring key releases by auto-repeat resolve the problem?

**Comment by shun-iwasawa** [comment]
@manongjohn 

> I've discovered why holding a tool shortcut for temporary access to that tool doesn't work in most cases.

Sorry for making a quick judgment but I just briefly tried the temporary switching of tools in my environment and the switching itself seems to work with no problem.
The issue #1100 and this seem to be not claiming that the temporary switching of tools itself is unavailable, but pointing about some inconveniences of this feature, right?

**Comment by manongjohn** [comment]
It appears due to how I was remotely using OT, it was doing something odd with my keystroke events and causing isAutoRepeat to always return false when holding 1 key. Oddly it correctly returned true if I held ctrl/shift+key. Why it was wrong in one scenario but not the other is a mystery.

Working directly on my machine, it now works correctly. There is no point for me to continue looking at this since I cannot duplicate it in a normal setting. Sorry for the misleading info. 

I did find that there are some open QTBUGs reported regarding isAutoRepeat not working properly in some cases so that may still be an issue.

**Comment by RodneyBaker** [comment]
This issue went back and forth quite a bit and may have gotten resolved in the meantime.
@tghart  Are you still experiencing the initial problem with using Spacebar and Geometry tool at the same time?

**Comment by RodneyBaker** [comment]
I believe this issue may have been resolved by denying ability to do both processes at same time.
It will be good to confirm any remaining problems.

**Comment by flurick** [comment]
> Hard coding space bar to do the same as the hand tool would work but the downside is that you couldn't change the shortcut.

Seems reasonable, if we use space bar as a modifier here it could still be mapped to action that do not require mouse drag, but all tools would work as expected.
(Quick hack [here](https://github.com/opentoonz/opentoonz/commit/710faa9ddc91f122c48f7248aca3005c527a066a))

**Comment by RodneyBaker** [comment]
We've lost track of the status of this problem relative to the original report and given the age of last update any clarification on current status (as well as any breakaway issues still outstanding) will be appreciated.

**Comment by RodneyBaker** [comment]
Clsoing as we can now assign Space key to shortcuts.
Please see:  #6199
